vault/programs/vault/src/lib.rs

```diff
@@ -0,0 +1,464 @@
+use anchor_lang::prelude::*;
+use anchor_lang::solana_program::{program::invoke, system_instruction};
+use anchor_spl::token::{self, Burn, CloseAccount, Mint, MintTo, Token, TokenAccount, Transfer};
+use rust_decimal::prelude::*;
+use std::ops::Deref;
+
+pub mod context;
+pub mod pyth_client;
+pub mod zeta_account;
+pub mod zeta_client;
+pub mod zeta_constants;
+pub mod zeta_context;
+pub mod zeta_utils;
+use crate::context::*;
+use crate::zeta_account::*;
+use crate::zeta_constants::*;
+use crate::zeta_utils::*;
+
+declare_id!("E1LM1zGDvJsEkbFJfLV2dyS83dfSAvbbfNG2YxSvuTz6");
+
+#[program]
+pub mod vault {
+
+    use super::*;
+
+    #[access_control(validate_epoch_times(epoch_times))]
+    pub fn initialize_vault(
+        ctx: Context<InitializeVault>,
+        vault_name: String,
+        vault_lamports: u64,
+        bumps: VaultBumps,
+        epoch_times: EpochTimes,
+    ) -> ProgramResult {
+        msg!("INITIALIZE vault");
+
+        let vault = &mut ctx.accounts.vault;
+
+        let name_bytes = vault_name.as_bytes();
+        let mut name_data = [b' '; 20];
+        name_data[..name_bytes.len()].copy_from_slice(name_bytes);
+
+        vault.vault_name = name_data;
+        vault.bumps = bumps;
+        vault.vault_authority = ctx.accounts.vault_authority.key();
+
+        vault.usdc_mint = ctx.accounts.usdc_mint.key();
+        vault.redeemable_mint = ctx.accounts.redeemable_mint.key();
+        vault.vault_usdc = ctx.accounts.vault_usdc.key();
+
+        vault.epoch_times = epoch_times;
+
+        // Transfer initial lamport balance to vault payer
+        msg!(
+            "Transferring {} lamports from `vault_authority` to `vault_payer`",
+            vault_lamports
+        );
+        invoke(
+            &system_instruction::transfer(
+                &ctx.accounts.vault_authority.key(),
+                &ctx.accounts.vault_payer.key(),
+                vault_lamports,
+            ),
+            &[
+                ctx.accounts.vault_authority.to_account_info(),
+                ctx.accounts.vault_payer.to_account_info(),
+                ctx.accounts.system_program.to_account_info(),
+            ],
+        )?;
+
+        Ok(())
+    }
+
+    #[access_control(deposit_withdraw_phase(&ctx.accounts.vault))]
+    pub fn init_user_redeemable(ctx: Context<InitUserRedeemable>) -> ProgramResult {
+        msg!("INIT USER REDEEMABLE");
+        Ok(())
+    }
+
+    #[access_control(deposit_withdraw_phase(&ctx.accounts.vault))]
+    pub fn exchange_usdc_for_redeemable(
+        ctx: Context<ExchangeUsdcForRedeemable>,
+        amount: u64,
+    ) -> ProgramResult {
+        msg!("EXCHANGE USDC FOR REDEEMABLE");
+        // While token::transfer will check this, we prefer a verbose err msg.
+        if ctx.accounts.user_usdc.amount < amount {
+            return Err(ErrorCode::LowUsdc.into());
+        }
+
+        // Transfer user's USDC to vault USDC account.
+        let cpi_accounts = Transfer {
```

Comment:
Can you use the into_context to move this logic outside of this body of code

Reply:
Code would just look nicer i.e. copy from flex snippet

```
        token::transfer(
            ctx.accounts
                .into_transfer_context()
                .with_signer(&[&vault_seeds[..]]),
            ctx.accounts.bid_vault.amount,
        )?;

        token::close_account(
            ctx.accounts
                .into_close_bid_vault_context()
                .with_signer(&[&vault_seeds[..]]),
        )?;
```

Reply:
Done! Looks lit now